src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,30 +1,93 @@
 package com.es.phoneshop.web;
 
-import com.es.phoneshop.model.exception.ProductNotFoundException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartService;
+import com.es.phoneshop.model.cart.HttpSessionCartService;
+import com.es.phoneshop.model.exception.OutOfStockException;
 import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
 import com.es.phoneshop.model.product.ProductDao;
+import com.es.phoneshop.model.recentlyViewed.HttpSessionViewedProductsService;
+import com.es.phoneshop.model.recentlyViewed.RecentlyViewedProducts;
+import com.es.phoneshop.model.recentlyViewed.ViewedProductsService;
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.Locale;
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ProductDao productDao;
+    private CartService cartService;
+    private ViewedProductsService viewedProductsService;
 
     @Override
     public void init(){
         productDao= ArrayListProductDao.getInstance();
+        cartService= HttpSessionCartService.getInstance();
+        viewedProductsService= HttpSessionViewedProductsService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException{
+        long productID=getProductID(request);
+        Product product=productDao.getProduct(productID);
+
+        RecentlyViewedProducts viewedProducts=viewedProductsService.getViewedProducts(request);
+        viewedProductsService.add(viewedProducts,product);
+        request.getSession().setAttribute("viewedProducts", viewedProducts);
+
+        showPage(request,response,product);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        long productID=getProductID(request);
+        Product product=productDao.getProduct(productID);
+
+        addToCart(request,response,product);
+    }
+
+    private void addToCart(HttpServletRequest request, HttpServletResponse response,Product product) throws ServletException, IOException {
+        String error =null;
```

Comment:
Here and in other places - coding style guides:

Locale locale=request.getLocale(); --> Locale locale = request.getLocale();
if (error!=null){ --> if (error != null) {

Style guides: https://google.github.io/styleguide/javaguide.html

Let IDEA format the file for you: https://www.jetbrains.com/help/idea/reformat-file-dialog.html

Ctrl + Shift + Alt + L or appropriate option in settings